[pytorch] Add return type for list of tensors

larryliu0820 opened the pull request:
Summary:
Adding codegen logic to return `at::TensorList` as a return type.

Although we don't really have a return type of `-> Tensor(a)[]` in `native_functions.yaml`, there are use cases for a `TensorList` return type.

If the type is a list and the element type is tensor, return a `at::TensorList`. Since the copy cost of an `ArrayRef` is very small, we don't need to return a reference.

Test Plan: Rely on unit test.

Reviewed By: iseeyuan

Differential Revision: D35737310

ezyang commented:
I kind of want to understand if this is just because this is the path of least resistance, or there really is an important reason to want to return a mutable list of tensors.

To explain, we have to first understand why inplace operations like neg_ have the signature `neg_(Tensor(a!) self) -> Tensor(a!)`. The return argument is guaranteed to be exactly the same as the input argument. So technically, the signature could also be `neg_(Tensor(a!) self) -> ()`. It is not expressed in this way for two reasons:

 * In the *user* facing API, inplace operations return themselves to make it easier to do method chaining, e.g., `neg_(...).add_(...)`
 * In TorchScript interpreter, we are operating a stack machine, and whether or not we "return" the argument says whether or not we pop the input argument off the stack, or keep it on the stack (so method chaining could compile to marginally more efficient code). This is very very marginal and I don't think it matters.

With lists of tensors, the method chaining justification doesn't apply at all. I don't think you have any interpreter-y things. So as far as I'm concerned, these functions can just return void. In fact, returning a list of tensors is problematic. If the input tensor list is not contiguous (as is the case for IValue tensor list), you have to reallocate it for output (incurring refcount bumps).

So yes, there is a bit of inconsistency if we return void. But I believe that we can make it work out without too much effort.

larryliu0820 committed:
[pytorch] Add return type for list of tensors (#76049)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76049

Adding codegen logic to return `at::TensorList` as a return type.

Although we don't really have a return type of `-> Tensor(a)[]` in `native_functions.yaml`, there are use cases for a `TensorList` return type.

If the type is a list and the element type is tensor, return a `at::TensorList`. Since the copy cost of an `ArrayRef` is very small, we don't need to return a reference.

Test Plan: Rely on unit test.

Reviewed By: iseeyuan

Differential Revision: D35737310

fbshipit-source-id: 94c7d36c7a52697d0f3b2736232b0fe8cdfd258e
